src/v/kafka/protocol/schemata/generator.py

```diff
@@ -1187,10 +1210,10 @@ class response;
 {{- struct_serde(field.type().value_type(), methods, "v") | indent }}
     return v;
 {%- else %}
-{%- set decoder, named_type = field.decoder(flex) %}
+{%- set decoder, named_type, has_primitives = field.decoder(flex) %}
 {%- if named_type == None %}
     return reader.{{ decoder }};
-{%- elif field.nullable() %}
+{%- elif field.nullable() and has_primitives == false%}
     {
         auto tmp = reader.{{ decoder }};
         if (tmp) {
```

Comment:
So the compile issue occurs when the Field is an array of primitive types AND is nullable. Thus the generator follows the highlighted branch (I call it the "null check" branch) when instead the "else" branch should be taken.

Checking if there are primitives solves the issue; however, I think returning a third value from the decoder method is poor style. What do you recommend @graphcareful ?

Reply:
leaving note, discussed offline